Utils/Dataflow/095_datasetInfoAMI/amiDatasets.py

```diff
@@ -153,9 +158,9 @@ def amiPhysValues(data):
         sys.stderr.write("(WARN) Could not process dataset '%s' due to "
                          "SSLError:%r\n" % (data['datasetname'], e))
         return False
-    except Exception:
-        sys.stderr.write("(WARN) No values found in AMI for dataset '%s'\n"
-                         % data['datasetname'])
+    except Exception as e:
+        sys.stderr.write("(WARN) Exception while processing dataset '%s':"
+                         "%r\n" % (data['datasetname'], e))
```

Comment:
Is it still "WARN", maybe it is more like an "ERROR"? I don\`t know for myself, in which situations we reach this point, so it is just a question.

"Exception while processing" sounds a bit... not friendly? I mean, error messages are supposed to be user-oriented and explain what happened, not simply state it. Maybe "Failed to process dataset", or "Dataset processing failure"?

Reply:
Done. Also changed the similar case catching SSL errors, for the same reason. Due to the changes the two cases (SSL errors and all other errors) perform the same actions, but this will change if I manage to transform the SSL errors into more meaningful form.

Reply:
Added traceback. Now this is more like it.

```
(ERROR) Failed to process dataset 'mc15_13TeV.342970.aMcAtNloPythia8EvtGen_A14NNPDF23LO_HZZvvqq900w15.recon.AOD.e4501_s2726_r7772_tid08345155_00': SSLError(336445449, u'[SSL] PEM lib (_ssl.c:2574)')
(DEBUG) Traceback start.
  File "./amiDatasets.py", line 165, in amiPhysValues
    format='json')
  File "/usr/lib/python2.7/site-packages/pyAMI/client.py", line 215, in execute
    self.httpClient.connect(endpoint)
  File "/usr/lib/python2.7/site-packages/pyAMI/httpclient.py", line 105, in connect
    context = context
  File "/usr/lib64/python2.7/httplib.py", line 1244, in __init__
    context.load_cert_chain(cert_file, key_file)
(DEBUG) Traceback end.
```

Reply:
Ok, thank you.

By the way, there\`s an implementation of traceback logging in `pyDKB` in [AbstractStage](https://github.com/PanDAWMS/dkb/blob/master/Utils/Dataflow/pyDKB/dataflow/stage/AbstractStage.py#L254) -- can\`t it be used here?

Reply:
Hm. It can be used, but I'm not sure that it should be.

`output_error()` is a method of `AbstractStage`, which in our case is represented by `stage` in `process()`. `amyPhysValues()` is called in `process()` with a single parameter `data`. This means that `amiPhysValues()` has no option to call `output_error()` in its current state, and `sys.exc_info()` call from `process()` does not catch the info from `amiPhysValues()`. Possibilities exist to make it work, so far I've thought of the following:

1. Pass `stage` as an argument to `amiPhysValues()`.
2. Return something (result of `sys.exc_info()`, for example) into `process` to work with.
3. Do 1 or 2, but with global variable or something else instead of argument/return.
4. Make (and use in this stage) a special class (`AMIStage` or something) based on `ProcessorStage` and turn `amiPhysValues()` into its method.

Neither of these look too appealing to me, more like unnecessary complications for the sake of using specific method to report an error. Perhaps 2 is the least "illogical", but we may need return status for something else.

Long story short, I think we should either process errors on the spot (and leave everything as-is), or, if we really need it, consider the implementation of some sort of general mechanism to return all errors (and not just SSLError, and not just in this stage) into process/stage and work with them there.

Reply:
> Possibilities exist to make it work, so far I've thought of the following:
> 1. Pass stage as an argument to amiPhysValues().
> 1. Return something (result of sys.exc_info(), for example) into process to work with.
> 1. Do 1 or 2, but with global variable or something else instead of argument/return.
> 1. Make (and use in this stage) a special class (AMIStage or something) based on ProcessorStage and turn amiPhysValues() into its method.

All the suggestions are equally bad, for my liking. Except for the last -- it could be a way to organize any of the stages, in fact. I mean -- to wrap the whole stage into an object with functions turned into methods, instead of `main()` plus the bunch of standalone functions. Doesn\`t look bad, it\`s just what we\`ve never done before.

However, there are many other possibilities -- and the most straightforward would be to simply allow the exception to pop from the `amiPhysValues()` and process it on the `process()` level. Why not? Besides, it would also allow more or less accurate handling of any other exception that might appear in this section:
https://github.com/PanDAWMS/dkb/blob/574ca5026f70c96783798712d787551745768c33/Utils/Dataflow/095_datasetInfoAMI/amiDatasets.py#L141-L143
It is, by the way, indeed what you\`re talking about:
> some sort of general mechanism to return all errors (and not just SSLError, and not just in this stage) into process/stage and work with them there.

-- and it is already invented for us.

Other option would be to look at the situation from the different point: not "how to apply what we have", but "how to make it comfortable". My thoughts are: if there *is* an implementation of what you need (and it includes some pretty stuff like proper prefixes for all the lines etc.), but hidden so deep that makes it difficult to use without acrobatic exercises -- it is a direct indicator that something\`s not right and worth improvement. Maybe not right now, but still deserves attention. So, from this point of view, the question is: how would you like it to be? what way you\`d feel natural to use the library implementation?

---

I\`m not saying anything of this should be done to make me accept this PR -- it can be done later by any of us, just with the same result; it\`s just "since we are here, let\`s not pass by this possibility of improvement and make life easier for those in the same situation -- should it be a month, or two (or more) from now".

I mean, it hit me because I personally spent some time figuring how to output the traceback in the common context of logging; here you did the same; next time it will be some one else, or one of us copy-and-pasting the solution invented before. Which is simply not right, and looks like an opportunity to learn how to do things better and avoid such situations in the future, when one can not use results of someone else's work and has to do it by him/herself.

Reply:
> However, there are many other possibilities -- and the most straightforward would be to simply allow the exception to pop from the amiPhysValues() and process it on the process() level.

Ah, you meant to use `raise`, right? Have not thought about that, and it seems to solve the problem in a proper way. Done: 8d1a11d.